CEED v4.0 release

v-dobrev opened the pull request:
Version 4.0 of the CEED package collection.

jedbrown committed:
petsc: add hip variant

jedbrown committed:
libceed: add 0.8, disable occa by default, and let autodetect AVX

Disabling OCCA because backend updates did not make this release and
there are some known bugs so most users won't have reason to use OCCA.

https://github.com/CEED/libCEED/pull/688

jedbrown committed:
WIP: ceed: 4.0 release

v-dobrev committed:
MFEM package updates (#19748)

* MFEM package updates

* mfem: flake8

* [mfem] Various fixes and tweaks.

[arpack-ng] Add a patch to fix building with IBM XL Fortran.

[libceed] Fix building with IBM XL C/C++.

[pumi] Add C++11 flag for version 2.2.3.

* [mfem] Fix the shared CUDA build.

Reported by: @MPhysXDev

* [mfem] Fix a TODO item

* [mfem] Tweak the AmgX dependencies

* [suite-sparse] Fix the version of the mpfr dependency

* MFEM: add initial HIP support using the ROCmPackage.

* MFEM: add 'slepc' variant.

* MFEM: update the patch for v4.2 for SLEPc.

* mfem: apply 'mfem-4.2-slepc.patch' just to v4.2.

v-dobrev committed:
ceed: apply 'spack style'

v-dobrev committed:
[mfem] Add a patch for mfem v4.2 to work with petsc v3.15.0.

[laghos] Add laghos version 3.1 based on the latest commit in
         the repository; this version works with mfem v4.2.

[ceed] For ceed v4.0 use laghos v3.1.

v-dobrev commented on var/spack/repos/builtin/packages/laghos/package.py:
```diff
@@ -21,6 +21,7 @@ class Laghos(MakefilePackage):
     maintainers = ['v-dobrev', 'tzanio', 'vladotomov']
 
     version('develop', branch='master')
+    version('3.1', commit='95f2b9d748146d94ab70086329a6c6f814767379')
```
@tzanio, @vladotomov, we should probably tag this commit with `v3.1` in the laghos repo. Adding this version was necessary since v3.0 does not work with mfem v4.2.

tzanio replied:
I am OK with that

v-dobrev replied:
Done.

v-dobrev committed:
[libceed] Explicitly set 'CC_VENDOR=icc' when using 'intel'
          compiler.

v-dobrev committed:
[mfem] Allow pumi >= 2.2.3 with mfem >= 4.2.0.

[ceed] Use pumi v2.2.5 with ceed v4.0.0.

v-dobrev committed:
[ceed] Explicitly use occa v1.1.0 with ceed v4.0.0.
       Use mfem@4.2.0+rocm with ceed@4.0.0+mfem+hip.

v-dobrev committed:
Merge branch 'develop' into packages/ceed-4.0

v-dobrev committed:
[ceed] Add NekRS v21 as a dependency for ceed v4.0.0.

v-dobrev committed:
[ceed] Fix NekRS version: 21 --> 21.0

v-dobrev commented:
Latest commit worked fine on Lassen with gcc v8.3.1 with
```console
./bin/spack install ceed+cuda ^magma cuda_arch=70 ^mfem cuda_arch=sm_70
```

v-dobrev committed:
[ceed] Propagate +cuda variant to petsc for ceed v4.0.

v-dobrev commented:
The NekRS package seems to have both `hip` (explicitly defined but not used) and `rocm` (inherited from `ROCmPackage`) variants which is redundant. Also, `depends_on('hip', when='+rocm')` is unnecessary since it is inherited from `ROCmPackage`, see https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/rocm.py.

Ping: @thilinarmtb.

On a similar topic: I think wee should add a `hip` (or `rocm`?) variant to `occa`. I also added relevant TODO comments about this in the `ceed` and `libceed` packages.

Ping: @dmed256.

v-dobrev committed:
[mfem] Propagate '+rocm' variant to some other packages.

v-dobrev committed:
[ceed] Use +rocm variant of nekrs instead of +hip.

thilinarmtb commented:
@v-dobrev : I will fix this in the nekRS package.

v-dobrev commented on var/spack/repos/builtin/packages/ceed/package.py:
```diff
@@ -125,6 +152,9 @@ class Ceed(BundlePackage):
     # MAGMA
+    # ceed-4.0
+    depends_on('magma@2.5.4', when='@4.0.0+cuda')
+    depends_on('magma@2.5.4+hip', when='@4.0.0+hip')
```
Currently the magma spack package does not have `hip` variant.

@stomov, @nbeams, is adding a `hip` variant for magma 2.5.4 possible at all? If so, is it relatively easy to add here?

v-dobrev replied:
ping: @stomov, @nbeams

nbeams replied:
There are not currently plans to add a `hip` spack variant for MAGMA 2.5.4, but there will be for the upcoming release of MAGMA 2.6.  Stan says he can hopefully work on the CMake and spack packages for 2.6 this week.

v-dobrev replied:
Okay, thanks for the update. Since we (most probably) don't want to wait for magma v2.6, I'll update the `ceed` package to remove comments about `magma+hip`. We can postpone the `magma+hip` support in `ceed` to the next CEED release.

nbeams replied:
I agree, that sounds like the best course for now.  Sorry about the delay.

v-dobrev committed:
[ceed] Do not enable magma with ceed@4.0.0+hip.

v-dobrev committed:
[libceed] Fix hip build with libceed@0.8.

v-dobrev committed:
[laghos] For v3.1, use the release .tar.gz file instead of commit.

v-dobrev commented:
> The NekRS package seems to have both `hip` (explicitly defined but not used) and `rocm` (inherited from `ROCmPackage`) variants which is redundant. Also, `depends_on('hip', when='+rocm')` is unnecessary since it is inherited from `ROCmPackage`, see https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/rocm.py.
> 
> Ping: @thilinarmtb.
> 
> On a similar topic: I think wee should add a `hip` (or `rocm`?) variant to `occa`. I also added relevant TODO comments about this in the `ceed` and `libceed` packages.
> 
> Ping: @dmed256.

Any progress on this?

thilinarmtb commented:
@v-dobrev : Do the following changes look okay ? If so, I will push them.

```diff
diff --git a/var/spack/repos/builtin/packages/nekrs/package.py b/var/spack/repos/builtin/packages/nekrs/package.py
index 731c34e21f..07b3f59c89 100644
--- a/var/spack/repos/builtin/packages/nekrs/package.py
+++ b/var/spack/repos/builtin/packages/nekrs/package.py
@@ -28,7 +28,7 @@ class Nekrs(Package, CudaPackage, ROCmPackage):
     variant('opencl',
             default=False,
             description='Activates support for OpenCL')
-    variant('hip',
+    variant('rocm',
             default=False,
             description='Activates support for HIP')

@@ -45,7 +45,6 @@ class Nekrs(Package, CudaPackage, ROCmPackage):
     depends_on('cmake')

     depends_on('cuda', when='+cuda')
-    depends_on('hip', when='+rocm')

     @run_before('install')
     def fortran_check(self):
```

v-dobrev commented:
@thilinarmtb, just remove the lines: https://github.com/spack/spack/blob/dc4f8578faaf08c1a1b1e3b553950d2c1efed759/var/spack/repos/builtin/packages/nekrs/package.py#L31-L33
The `rocm` variant is already defined by the `ROCmPackage`: https://github.com/spack/spack/blob/6ecec24105269ae3618ae0d2a52e2de9662e2c56/lib/spack/spack/build_systems/rocm.py#L99

Removing the dependence on `hip`: https://github.com/spack/spack/blob/dc4f8578faaf08c1a1b1e3b553950d2c1efed759/var/spack/repos/builtin/packages/nekrs/package.py#L48
is correct because it is already in `ROCmPackage`: https://github.com/spack/spack/blob/6ecec24105269ae3618ae0d2a52e2de9662e2c56/lib/spack/spack/build_systems/rocm.py#L108

v-dobrev commented:
@thilinarmtb, by the way, similar comments apply to the `cuda` variant: https://github.com/spack/spack/blob/dc4f8578faaf08c1a1b1e3b553950d2c1efed759/var/spack/repos/builtin/packages/nekrs/package.py#L25-L27
and the `cuda` dependency line: https://github.com/spack/spack/blob/dc4f8578faaf08c1a1b1e3b553950d2c1efed759/var/spack/repos/builtin/packages/nekrs/package.py#L47
These are already defined in the `CudaPackage`, here:
https://github.com/spack/spack/blob/6ecec24105269ae3618ae0d2a52e2de9662e2c56/lib/spack/spack/build_systems/cuda.py#L34-L35
and here: https://github.com/spack/spack/blob/6ecec24105269ae3618ae0d2a52e2de9662e2c56/lib/spack/spack/build_systems/cuda.py#L49

thilinarmtb committed:
Remove cuda & hip variants as they are inherited

thilinarmtb commented:
@v-dobrev : I pushed the changes after testing on Summit.

v-dobrev committed:
[ceed] Remove comments and FIXMEs about 'magma+hip'.

v-dobrev committed:
[ceed] [libceed] Remove TODOs about occa + hip.

v-dobrev commented:
TODO:
- [x] In `petsc`, `libceed`, and `ceed`: replace the `+hip` variant with `+rocm` by inheriting from `ROCmPackage`.
- [x] In `mfem`: replace the dependency `'libceed+hip'` with `'libceed+rocm'.`

jedbrown committed:
Merge remote-tracking branch 'origin/develop' into packages/ceed-4.0

jedbrown committed:
libceed: use ROCmPackage and +rocm

jedbrown committed:
petsc: use ROCmPackage for HIP

jedbrown committed:
libceed, petsc: use CudaPackage

jedbrown committed:
ceed: forward cuda_arch and amdgpu_target

jedbrown commented on var/spack/repos/builtin/packages/ceed/package.py:
```diff
@@ -43,9 +39,13 @@ class Ceed(BundlePackage):
+    depends_on('libceed@0.8~rocm', when='@4.0.0~rocm')
+    for target in ROCmPackage.amdgpu_targets:
+        depends_on('libceed@0.8+rocm amdgpu_target={0}'.format(target),
+                   when='@4.0.0+rocm amdgpu_target={0}'.format(target))
```
@becker33 Is this what you meant when you said (on Slack) that the variant needs to be forwarded to each dependent package? The spec makes sense with this change, though it's tedious going and all the BundlePackages would need this treatment.

v-dobrev committed:
[mfem] Use Spack's CudaPackage as a base class; as a result,
       'cuda_arch' values should not include the 'sm_' prefix.
       Also, propagate 'cuda_arch' and 'amdgpu_target' variants
       to enabled dependencies.

jedbrown commented on var/spack/repos/builtin/packages/mfem/package.py:
```diff
@@ -438,7 +451,7 @@ def find_optional_library(name, prefix):
             if '+cuda' in spec:
                 cxxflags += [
-                    '-x=cu --expt-extended-lambda -arch=%s' % cuda_arch,
+                    '-x=cu --expt-extended-lambda -arch=sm_%s' % cuda_arch,
```
I think Todd describes this flag as being handled automatically by the compiler wrappers when using `CudaPackage`.

v-dobrev replied:
This is a flag for `nvcc` -- I don't think spack has a wrapper for that -- unless that is a recent addition.

v-dobrev commented:
The following install worked fine on Lassen (with gcc-8.3.1 and various other customizations in `compilers.yaml` and `packages.yaml`):
```console
./bin/spack install ceed+cuda cuda_arch=70
```

jedbrown committed:
petsc: variant is +rocm, package name is hip

jedbrown reviewed (dismissed):
LGTM. @balay may want to review the PETSc changes.

tzanio reviewed (dismissed):
Thank you @v-dobrev, @jedbrown and @thilinarmtb !

tzanio commented:
We plan to merge this on Wednesday (06/02/21). Please speak up if you have feedback or concerns.

tzanio commented:
@jedbrown, can you resolve the conflict in petsc so we can merge?

jedbrown committed:
Merge remote-tracking branch 'origin/develop' into packages/ceed-4.0

Resolve petsc conflict on new hwloc and openmp variants.

jedbrown reviewed (approved):
This should be good to merge once tests pass.
